### PR TITLE
libecpint integration

### LINKED ISSUE #567: ECP integrals via library
Replace our implementation by https://github.com/robashaw/libecpint ?

### PR DESCRIPTION
**Pull request by junghans**
Ultimately fix #567

### CODE REVIEW

**Comment by junghans** [comment]
Let's see if we have it everywhere.

**Comment by junghans** [comment]
Works now, @rubengerritsen @JensWehner I leave the actual work for you guys!

**Comment by JensWehner** [comment]
@junghans Thanks for the help. We will make it work. :)

**Comment by JensWehner** [comment]
@votca-bot changelog: Integrade libecpint library.

**Comment by JensWehner** [comment]
@votca-bot changelog: use libecpint library for pseudo potentials

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by junghans** [comment]
@votca-bot copyright

**Comment by JensWehner** [comment]
@votca-bot changelog: using libecpint for pseudo potentials

**Comment by JensWehner** [comment]
this works locally but atm is blocked by https://github.com/robashaw/libecpint/pull/26

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot changelog using libecpint for pseudo potentials

**Comment by JensWehner** [comment]
ahh results are equivalent to orca to 0.1mHrt

**Comment by junghans** [comment]
> @votca-bot changelog using libecpint for pseudo potentials

missing ":" after changelog.

**Comment by junghans** [comment]
@votca-bot changelog: using libecpint for pseudo potentials

**Comment by junghans** [comment]
CI error:
```
2021-01-16T16:22:43.7296708Z /usr/lib64/ccache/clang++ -DDEBUG -Dvotca_xtp_EXPORTS -Ixtp/src/libxtp -I/__w/xtp/xtp/votca/xtp/src/libxtp -I/__w/xtp/xtp/votca/xtp/include -Ixtp/include -Ixtp/include/votca/xtp -I/__w/xtp/xtp/votca/csg/include -I/__w/xtp/xtp/votca/tools/include -Itools/include -Itools/include/votca/tools -isystem /usr/include/eigen3 -isystem /usr/include/libint2 -isystem /usr/include/libecpint -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Wno-sign-conversion -Werror -Wno-deprecated-copy -g -fPIC -fopenmp=libomp -std=c++14 -MD -MT xtp/src/libxtp/CMakeFiles/votca_xtp.dir/ecpaobasis.cc.o -MF xtp/src/libxtp/CMakeFiles/votca_xtp.dir/ecpaobasis.cc.o.d -o xtp/src/libxtp/CMakeFiles/votca_xtp.dir/ecpaobasis.cc.o -c /__w/xtp/xtp/votca/xtp/src/libxtp/ecpaobasis.cc
2021-01-16T16:22:43.7304572Z /__w/xtp/xtp/votca/xtp/src/libxtp/ecpaobasis.cc:134:18: error: variable 'I' is uninitialized when passed as a const reference argument here [-Werror,-Wuninitialized-const-reference]
2021-01-16T16:22:43.7305784Z     table.addCol(I, "atomid", HOFFSET(data, atomid));
2021-01-16T16:22:43.7306225Z                  ^
2021-01-16T16:22:43.7307533Z /__w/xtp/xtp/votca/xtp/src/libxtp/ecpaobasis.cc:135:18: error: variable 'd' is uninitialized when passed as a const reference argument here [-Werror,-Wuninitialized-const-reference]
2021-01-16T16:22:43.7308932Z     table.addCol(d, "posX", HOFFSET(data, x));
2021-01-16T16:22:43.7309329Z                  ^
2021-01-16T16:22:43.7309671Z 2 errors generated.
```

**Comment by junghans** [comment]
@votca-bot changelog: using libecpint for pseudo potentials

**Comment by junghans** [comment]
CI error:
```
2021-01-16T17:29:32.5733942Z [ 85%] Linking CXX executable unit_test_ecpaobasis
2021-01-16T17:29:32.5819162Z cd /__w/xtp/xtp/builddir/xtp/src/tests && /usr/bin/cmake -E cmake_link_script CMakeFiles/unit_test_ecpaobasis.dir/link.txt --verbose=1
2021-01-16T17:29:32.6057864Z /usr/lib64/ccache/g++  -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Werror -Wno-deprecated-copy -O2 -DNDEBUG CMakeFiles/unit_test_ecpaobasis.dir/test_ecpaobasis.cc.o -o unit_test_ecpaobasis  -Wl,-rpath,/__w/xtp/xtp/builddir/xtp/src/libxtp:/__w/xtp/xtp/builddir/csg/src/libcsg:/__w/xtp/xtp/builddir/tools/src/libtools ../libxtp/libvotca_xtp.so.2022 /usr/lib64/libboost_unit_test_framework.so ../../../csg/src/libcsg/libvotca_csg.so.2022 ../../../tools/src/libtools/libvotca_tools.so.2022 /usr/lib64/libboost_program_options.so /usr/lib64/libfftw3.so /usr/lib64/libhdf5_cpp.so /usr/lib64/libhdf5.so /usr/lib/gcc/x86_64-redhat-linux/10/libgomp.so /usr/lib64/libpthread.so /usr/lib64/libint2.so 
2021-01-16T17:29:32.7749448Z /usr/bin/ld: CMakeFiles/unit_test_ecpaobasis.dir/test_ecpaobasis.cc.o: undefined reference to symbol '_ZN9libecpint3ECPC1ERKS0_'
2021-01-16T17:29:32.7750667Z /usr/bin/ld: /usr/lib64/libecpint.so.1: error adding symbols: DSO missing from command line
2021-01-16T17:29:32.7767296Z collect2: error: ld returned 1 exit status
```
let me try to fix that.

**Comment by junghans** [comment]
Now the test fail:
```
2021-01-16T18:42:42.1506284Z 10/83 Test #196: unit_test_aopotential .....................***Exception: SegFault  0.37 sec
2021-01-16T18:42:42.1506846Z munmap_chunk(): invalid pointer
2021-01-16T18:42:42.1507211Z Running 3 test cases...
2021-01-16T18:42:42.1508502Z unknown location(0): [4;31;49mfatal error: in "aopotential_test/aopotentials_test": signal: SIGABRT (application abort requested)[0;39;49m
2021-01-16T18:42:42.1509462Z /__w/xtp/xtp/votca/xtp/src/tests/test_aopotential.cc(50): [1;36;49mlast checkpoint[0;39;49m
2021-01-16T18:42:42.1510501Z unknown location(0): [4;31;49mfatal error: in "aopotential_test/large_l_test": memory access violation at address: 0x00000000: no mapping at fault address[0;39;49m
2021-01-16T18:42:42.1511466Z /__w/xtp/xtp/votca/xtp/src/tests/test_aopotential.cc(275): [1;36;49mlast checkpoint[0;39;49m
2021-01-16T18:42:42.9549783Z       Start 207: unit_test_ecpaobasis
2021-01-16T18:42:43.0346763Z 21/83 Test #207: unit_test_ecpaobasis ......................***Failed    0.08 sec
2021-01-16T18:42:43.0347299Z munmap_chunk(): invalid pointer
2021-01-16T18:42:43.0347722Z Running 1 test case...
2021-01-16T18:42:43.0349296Z unknown location(0): [4;31;49mfatal error: in "ecpaobasis_test/Serializing": signal: SIGABRT (application abort requested)[0;39;49m
2021-01-16T18:42:43.0350566Z /__w/xtp/xtp/votca/xtp/src/tests/test_ecpaobasis.cc(39): [1;36;49mlast checkpoint: "Serializing" test entry[0;39;49m
2021-01-16T18:42:43.0351022Z 
2021-01-16T18:42:43.0351563Z [1;31;49m*** 1 failure is detected in the test module "ecpaobasis_test"
2021-01-16T18:42:43.0352066Z [0;39;49m
```

**Comment by JensWehner** [comment]
I think that is the libecpint memory leak

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@rubengerritsen As the new libecpint release is up, this can be merged soon. Do you want to review?

**Comment by JensWehner** [comment]
@votca-bot format

File: CMakeLists.txt [hunk 1]
```diff
@@ -50,6 +50,13 @@ pkg_check_modules(LIBXC REQUIRED IMPORTED_TARGET libxc)
 if(NOT TARGET PkgConfig::LIBINT)
   pkg_check_modules(LIBINT REQUIRED IMPORTED_TARGET libint2>=2.6)
 endif()
+find_package(ecpint REQUIRED)
+set_package_properties(ecpint PROPERTIES TYPE REQUIRED PURPOSE "Calculate pseudo potentials")
+if(NOT TARGET ECPINT::ecpint) # newer versions are namespaced.
+  set_target_properties(ecpint PROPERTIES IMPORTED_GLOBAL TRUE)
+  add_library(ECPINT::ecpint ALIAS ecpint)
+endif()
+message(STATUS "Found libecpint: ECP${ecpint_DIR}")
```
Comment (junghans) [inline CMakeLists.txt:59]:
```suggestion
message(STATUS "Found libecpint: ${ecpint_DIR}")
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -50,6 +50,13 @@ pkg_check_modules(LIBXC REQUIRED IMPORTED_TARGET libxc)
 if(NOT TARGET PkgConfig::LIBINT)
   pkg_check_modules(LIBINT REQUIRED IMPORTED_TARGET libint2>=2.6)
 endif()
+find_package(ecpint REQUIRED)
+set_package_properties(ecpint PROPERTIES TYPE REQUIRED PURPOSE "Calculate pseudo potentials")
+if(NOT TARGET ECPINT::ecpint) # newer versions are namespaced.
+  set_target_properties(ecpint PROPERTIES IMPORTED_GLOBAL TRUE)
+  add_library(ECPINT::ecpint ALIAS ecpint)
+endif()
```
Comment (junghans) [inline CMakeLists.txt:58]:
```suggestion
```

Not needed anymore.

File: CMakeLists.txt [hunk 3]
```diff
@@ -50,6 +50,13 @@ pkg_check_modules(LIBXC REQUIRED IMPORTED_TARGET libxc)
 if(NOT TARGET PkgConfig::LIBINT)
   pkg_check_modules(LIBINT REQUIRED IMPORTED_TARGET libint2>=2.6)
 endif()
+find_package(ecpint REQUIRED)
```
Comment (junghans) [inline CMakeLists.txt:53]: Not right now, but later (next week?), we should required version 1.5 here.

Comment (JensWehner) [reply CMakeLists.txt:53]: 1.0.5? Actually we should require it now, because all the tests segfault otherwise.

Comment (junghans) [reply CMakeLists.txt:53]: Well, fedora and opensuse didn't receive the 1.0.5 package yet. https://bodhi.fedoraproject.org/updates/FEDORA-2021-5464f657f7 will need 7 more days.

Comment (junghans) [reply CMakeLists.txt:53]: But Fedora has 1.0.4 with your patch.

Comment (junghans) [reply CMakeLists.txt:53]: Ok, there is another bug in libecpint, that prevents from exporting the right version in the cmake target, so there is nothing we can right now.

Comment (junghans) [reply CMakeLists.txt:53]: We can required 1.0.6 once https://github.com/robashaw/libecpint/pull/32 is through. ;-)

Comment (JensWehner) [reply CMakeLists.txt:53]: okay

**Comment by JensWehner** [comment]
@rubengerritsen this can be reviewed. :)

**Review by rubengerritsen — APPROVED** [review]

**Pull request merged**